src/NLog/LayoutRenderers/DateLayoutRenderer.cs

```diff
@@ -74,21 +75,69 @@ public DateLayoutRenderer()
+        private System.Collections.Generic.KeyValuePair<DateTime, string> cachedUtcTime = new System.Collections.Generic.KeyValuePair<DateTime, string>();
+        private System.Collections.Generic.KeyValuePair<DateTime, string> cachedLocalTime = new System.Collections.Generic.KeyValuePair<DateTime, string>();
+
         /// <summary>
         /// Renders the current date and appends it to the specified <see cref="StringBuilder" />.
         /// </summary>
         /// <param name="builder">The <see cref="StringBuilder"/> to append the rendered data to.</param>
         /// <param name="logEvent">Logging event.</param>
         protected override void Append(StringBuilder builder, LogEventInfo logEvent)
         {
+            var formatProvider = GetFormatProvider(logEvent, Culture);
+
             var ts = logEvent.TimeStamp;
             if (this.UniversalTime)
             {
                 ts = ts.ToUniversalTime();
+                AppendDateLayout(builder, formatProvider, ts, ref cachedUtcTime);
+            }
+            else
+            {
+                AppendDateLayout(builder, formatProvider, ts, ref cachedLocalTime);
             }
+        }
 
-            var formatProvider = GetFormatProvider(logEvent, Culture);
-            builder.Append(ts.ToString(this.Format, formatProvider));
+        private static bool IsLowTimeResolutionLayout(string dateTimeFormat)
+        {
+            for (int i = 0; i < lowTimeResolutionChars.Length; ++i)
+                dateTimeFormat = dateTimeFormat.Replace(lowTimeResolutionChars[i], ' ');
+            for (int i = 0; i < dateTimeFormat.Length; ++i)
+            {
+                if (char.IsLetter(dateTimeFormat[i]))
+                    return false;
+            }
+            return true;
+        }
+
+        private void AppendDateLayout(StringBuilder builder, IFormatProvider formatProvider, DateTime timestamp, ref System.Collections.Generic.KeyValuePair<DateTime, string> cachedTime)
+        {
+            if (ReferenceEquals(formatProvider, CultureInfo.InvariantCulture))
+            {
+                if (cachedTime.Key != DateTime.MinValue)
+                {
+                    if (cachedTime.Key == timestamp.Date.AddHours(timestamp.Hour))
+                    {
+                        builder.Append(cachedTime.Value);
+                        return; // Cache hit
+                    }
+                }
+                else if (cachedTime.Value == null)
+                {
+                    // Check if caching should be used
+                    if (IsLowTimeResolutionLayout(this.Format))
+                        cachedTime = new System.Collections.Generic.KeyValuePair<DateTime, string>(DateTime.MaxValue, string.Empty);
+                    else
+                        cachedTime = new System.Collections.Generic.KeyValuePair<DateTime, string>(DateTime.MinValue, string.Empty);
```

Comment:
Could you explain me why we would store `DateTime.MaxValue`/`DateTime.MinValue`? What it's saying?

Also would be nice to extract the datetime to a variable to remove the last code duplication.